README.md
<p align='center'>
  <p align='center'><img width='150' src='https://raw.githubusercontent.com/coding-show/dort/master/client/static/images/favicon.png' /></p>
  <p align='center'>
    File sharing through web with wonderful alternation.</br>
    文件演示分享Web工具
  </p>
</p>

### Introduction - ([https://coding-tool.github.io/dort/](https://coding-tool.github.io/dort/))
Dort is a npm tool to share your editing, file structure through web during presentation or code review.

Dort 是一个使用者在 **讲演** 或者 **代码审阅** 过程 分享 **项目文件**、**编辑过程** 的web工具。

Everybody in the same network can browse through the files, check file history, copy the code, download the project, preview images, etc.

同一内网中的用户可以 **浏览文件**，**查看文件编辑版本**，**复制文本**，**下载整个项目**，**查看图片**，甚至可以与所有打开web的用户进行交流。

### Feature
- Easy for Client (all platform with browser) 便于使用
- Live Update 实时刷新
- Low Network Traffic 低网络负担

### Demo
![](https://github.com/coding-show/dort/blob/master/media/screenshot1.png?raw=true)
![](https://github.com/coding-show/dort/blob/master/media/screenshot2.png?raw=true)
![](https://github.com/coding-show/dort/blob/master/media/screenshot3.png?raw=true)

### Instruction
1. Prepare NodeJS and NPM environment 准备nodejs和npm环境

2. Install package 安装包
``` bash
# npm install -g dort
```

3. Launch dort inside your project folder 启动server端
```
# cd /your-files
# dort
```

4. Open default browser with default port - [http://localhost:4574](http://localhost:4574) 端口可设置

5. Shre the url to any body in the same network 其他内网用户可以通过相应端口访问

### Configuration
params | Type | Default	| Description
-------- | ---- | ------- | -----------
debug | number	| 1	| develop levels
path	| string	| ''	| path of file you want open
name	| string	| ''	| name of project
port	| number	| 4574	| port for web
exclude	| array	| ['**/.DS_Store', '**/.log', ...]	| file to exclude
test	| bool	| false	| Enable test mode or not
max-file-size	| number	| 99999	| max file size

Then you can launch dort as
``` bash
# dort "My project" --debug 0 --path /file-path --port 1234 --exclude "node_modules/**" --test true --max-file-size 99999
# // or
# dort "My project" -d 0 -pa /file-path -po 1234 -e "node_modules/**" -t true -m 99999
```


### TODO
- [x] node procedure to watch files in local
- [x] state management
- [x] FE web
- [x] chat and barrage
- [ ] File content interactive


CHANGELOG.md
## [1.1.0](https://github.com/coding-tool/dort/releases/tag/v1.1.0)

### Features
- [x] node procedure to watch files in local
- [x] state management
- [x] FE web
- [x] chat and barrage


lib/index.js
/**
 * client端入口文件
 */

require('colors')
const ip = require('ip')
const socketIo = require('socket.io')

const Watch = require('./watch')
const Site = require('./site.js')
const Projects = require('./projects.js')

const packageConfig = require('../package.json')

class App {
  constructor () {
    console.log('---'.rainbow + ' Dort '.bold + packageConfig.version.bold + ' ----------------------'.rainbow)

    // 导入配置参数
    this.setConfig()

    // 设置site
    this.setSite(
      () => {
        this.setSocket()
        this.setProjects()
      },
      () => {
        console.log('[dort]'.green.bold + ' - ' + 'seems to be running already'.cyan)
      },
      () => {
        console.log('[dort]'.green.bold + ' - ' + 'error'.red)
      }
    )

    // 开启监控模块
    this.setWatch()
  }

  // 解析命令行参数
  setConfig () {
    const config = {}

    const args = require('yargs')
      .option('debug', {
        alias: 'd',
        describe: '开发调试等级，用来规范打印信息',
        default: 1,
        type: 'number'
      })
      .option('path', {
        alias: 'pa',
        describe: '目录地址',
        default: '',
        type: 'string'
      })
      .option('name', {
        alias: 'n',
        describe: '工程名称',
        default: '',
        type: 'string'
      })
      .option('port', {
        alias: 'po',
        describe: '服务端口',
        default: 4574,
        type: 'number'
      })
      .option('exclude', {
        alias: 'e',
        describe: '需要排除掉的文件',
        default:
        [
          '**/.DS_Store',
          '**/node_modules/**',
          '**/vendor/**',
          '**/.git',
          '**/.vscode',
          '**/.env',
          '**/.log',
          '.idea/**',
          '**/*___jb_old___',
          '**/*___jb_tmp___'
        ],
        type: 'array'
      })
      .option('test', {
        alias: 't',
        describe: '开启测试工程',
        default: false,
        type: 'boolean'
      })
      .option('max-file-size', {
        alias: 'm',
        describe: '文件大小限制',
        default: 99999,
        type: 'number'
      })
      .argv

    config.debug = args.debug
    config.path = args.path
    config.name = args.name
    config.port = args.port
    config.exclude = args.exclude
    config.test = args.test

    // 工程名单独处理，可以是第一个arguments参数
    if (config.name === '') {
      if (args._.length > 0 && typeof args._[0] === 'string' && args._[0] !== '') {
        config.name = args._[0]
      } else {
        config.name = 'DEFAULT'
      }
    }

    // 定义domain
    config.domain = `http://${ip.address()}${config.port !== 80 ? ':' + config.port : ''}`

    this.config = config

    if (this.config.debug >= 1) {
      for (const _configKey in this.config) {
        const _configValue = '' + this.config[_configKey]
        console.log('[dort]'.yellow.bold + ' - ' + '<' + _configKey.green + '>' + ': ' + _configValue.cyan)
      }
    }
  }

  setSite (_success, _isRunning, _error) {
    this.site = new Site(this.config, _success, _isRunning, _error)
  }

  setSocket () {
    this.config.socket = socketIo.listen(this.site.server)

    // 新建socket监听watchSocket
    const watchSocket = this.config.socket.of('/watch')

    watchSocket.on('connection', socket => {
      let project

      if (this.config.debug >= 1) {
        console.log('[dort]'.green.bold + ' - ' + 'socket connect'.cyan + ' - ' + socket.id.cyan)
      }

      socket.on('start_project', data => {
        project = this.projects.createProject(data)
        this.site.createProjectRoute(project)

        if (this.config.debug >= 1) {
          console.log('[dort] > socket'.green.bold + ' - ' + 'start_project'.cyan)
        }
      })

      socket.on('create_file', data => {
        project.files.create(data.path, data.content)
        project.isUpdate = true

        if (this.config.debug >= 1) {
          console.log('[dort] > socket'.green.bold + ' - ' + 'create_file'.cyan + ' - ' + data.path.cyan)
        }
      })

      socket.on('update_file', data => {
        project.files.createVersion(data.path, data.content)
        project.isUpdate = true

        if (this.config.debug >= 1) {
          console.log('[dort] > socket'.green.bold + ' - ' + 'update_file'.cyan + ' - ' + data.path.cyan)
        }
      })

      socket.on('delete_file', data => {
        project.files.delete(data.path)
        project.isUpdate = true

        if (this.config.debug >= 1) {
          console.log('[dort] > socket'.green.bold + ' - ' + 'delete_file'.cyan + ' - ' + data.path.cyan)
        }
      })

      socket.on('disconnect', () => {
        this.projects.deleteProject(project.slug)

        if (this.config.debug >= 1) {
          console.log('[dort] > socket'.green.bold + ' - ' + 'disconnect'.cyan + ' - ' + socket.id.cyan)
        }
      })
    })
  }

  setProjects () {
    this.projects = new Projects(this.config)
  }

  setWatch () {
    this.watch = new Watch(this.config)
  }
}

module.exports = App
